File: 11.PrefixSum前缀和/leetcode_2906_constructProductMatrix.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> constructProductMatrix(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        int modNum = 12345;
        vector<vector<int>> res(m, vector<int>(n, 1));
        int prod = 1;
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                res[i][j] = (res[i][j]*prod)%modNum;
                prod = (1ll*grid[i][j]*prod)%modNum;
            }
        }

        prod = 1;
        for (int i=m-1; i>=0; i--){
            for (int j=n-1; j>=0; j--){
                res[i][j] = (res[i][j]*prod)%modNum;
                prod = (1ll*prod*grid[i][j])%modNum;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<vector<int>> grid = {{1,2},{3,4}};
    vector<vector<int>> res = s.constructProductMatrix(grid);
    for (auto& row : res){
        for (auto& i : row){
            cout << i << " ";
        }
        cout << endl;
    }
}

File: 11.PrefixSum前缀和/leetcode_2906_constructProductMatrix.py
from typing import List

class Solution:
    def constructProductMatrix(self, grid: List[List[int]]) -> List[List[int]]:
        modNum = 12345
        m, n = len(grid), len(grid[0])
        preProd = [1]*(m*n+1)
        sufProd = [1]*(m*n+1)
        for i in range(m):
            for j in range(n):
                indx = i*n+j
                x = grid[i][j]
                preProd[indx+1] = (preProd[indx]*x)%modNum

                indx2 = (m-i-1)*n+(n-j-1)
                x2 = grid[m-i-1][n-j-1]
                sufProd[indx2] = (sufProd[indx2+1]*x2)%modNum
        
        res = [[0]*n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                indx = i*n+j
                res[i][j] = (preProd[indx]*sufProd[indx+1])%modNum
        return res

if __name__ == "__main__":
    s = Solution()
    grid = [[1,2],[3,4]]
    print(s.constructProductMatrix(grid))
